src/services/qix/session.js

```diff
@@ -133,6 +170,143 @@ class Session {
+
+        const handle = response.result.qReturn.qHandle;
+        app.api.handle = handle;
+        restoredApis.add(handle, app.api);
+        return this.Promise.resolve(app.api);
+      });
+  }
+
+  /**
+  * Function used to restore the APIs on the doc.
+  * @param {Object} doc - The doc API on which the APIs we want to restore exist.
+  * @param {Object} restoredApis - The API-cache containing restored APIs.
+  * @param {Array} closed - A list of APIs that has been closed when resuming the session.
+  * @returns {Object} Returns a promise instance.
+  */
+  restoreDocObjects(doc, restoredApis, closed) {
+    const tasks = [];
+
+    const entries = this.apis.getApis().filter(entry => entry.api.type !== 'Global' && entry.api.type !== 'Doc');
+
+    entries.forEach((entry) => {
+      const api = entry.api;
+      const method = api.getMethodName;
+
+      if (!method) {
+        closed.push(api);
+      } else {
+        const request = this.rpc.send({
+          method,
+          handle: doc.handle,
+          params: [api.id],
+        })
+        .then((response) => {
+          if (response.error || !response.result.qReturn.qHandle) {
+            closed.push(api);
+          } else {
+            api.handle = response.result.qReturn.qHandle;
+            restoredApis.add(api.handle, api);
+          }
+        });
+        tasks.push(request);
+      }
+    });
+    return Promise.all(tasks);
+  }
+
+  /**
+  * Function used to resume the session
+  * @param {Boolean} onlyIfAttached - if true, resume only if the session was re-attached.
+  * @returns {Object} Returns a promise instance.
+  */
+  resume(onlyIfAttached) {
+    if (!this.suspended) {
+      return this.Promise.reject(new Error('Not suspended'));
+    }
+
+    this.unregisterRpcListeners();
+    const restoredApis = new ApiCache();
+    const closed = [];
+
+    return this.rpc.reopen(ON_ATTACHED_TIMEOUT)
+      .then((sessionState) => {
+        if (sessionState === 'SESSION_CREATED' && onlyIfAttached) {
+          return this.Promise.reject(new Error('Not attached'));
+        }
+        return this.Promise.resolve(sessionState);
+      })
+      .then(sessionState => this.restoreGlobal(restoredApis)
+        .then(() => this.restoreDoc(sessionState, restoredApis))
+        .then((doc) => {
+          if (doc) {
+            return this.restoreDocObjects(doc, restoredApis, closed);
+          }
+          return this.Promise.resolve();
+        })
+      )
+      .then(() => {
+        this.apis = restoredApis;
+        this.suspended = false;
+        this.registerRpcListeners();
+        closed.forEach(api => api.emit('closed'));
+        this.apis.getApis().filter(entry => entry.api.type !== 'Global').forEach((entry) => {
+          entry.api.emit('changed');
+        });
+        this.emit('resumed');
+      })
+      .catch(err => this.rpc.close().then(() => {
+        this.registerRpcListeners();
+        return this.Promise.reject(err);
+      }));
+  }
```

Comment:
In general, the `resume()` function is really complex, could we move out more logic from it to smaller functions instead of inline anonymous functions?